server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -543,7 +547,14 @@ private static Map<String, Mapper> buildMergedMappers(
                 Mapper mergeIntoMapper = mergedMappers.get(mergeWithMapper.simpleName());
 
                 if (mergeIntoMapper == null) {
-                    mergedMappers.put(mergeWithMapper.simpleName(), mergeWithMapper);
+                    if (mergeWithMapper instanceof ObjectMapper om) {
+                        addNewObjectMapper(reason, objectMergeContext, mergedMappers, om);
+                    } else {
+                        objectMergeContext.addFieldIfPossible(
+                            mergeWithMapper,
+                            m -> mergedMappers.put(mergeWithMapper.simpleName(), mergeWithMapper)
```

Comment:
if you keep the consumer, it may make sense to use `m` instead of `mergeWithMapper`, although they are in practice the same. Maybe using a consumer is flawed :)